apis.md

```diff
@@ -9,3 +9,10 @@ The following OpenTracing APIs are currently available in the RFC stage:
 
 Please refer to the README files in the respective projects for examples of usage.
 
+### Implementation Notes
+
+OpenTracing APIs intend to express a similar experience to the user, while still being idiomatic to the language or framework. The following concepts apply to all variants.
+
+A **Span** is shared mutable state. It accumulates tags and logs, and can also create children. Finishing a span means capturing its state as a SpanRecord.
+
+A **SpanRecord** is an immutable representation of a completed span. It can be reported directly to an OpenTracing system, or encoded into a representation of an alternate system, such as HTrace or Zipkin.
```

Comment:
> IOTW, at least in api-java I plan to use the same route as we do in zipkin for span stores. Which is to define tests and a fake that live in the authoritative repo. Downstream implementations must pass or explicitly skip these tests. This simplifies answer to the question of what behavior should be vs what one or another downstream do.

@adriancole yes for langs like Java/Go/etc it makes more sense to define a fake. Not sure if you saw the test harness I added in Python, it basically exercises all API methods against a given tracer, and in the main api-python repo it just runs against the no-op tracer which itself serves as the API definition.

Unfortunately, given that the API is mostly "write", there are limited things we can do in the tests to actually check correctness of the implementation. One possibility is to define some "read" API that must be implemented by the Reporter or some anonymous component of the tracer based on the data it captured, purely for compliance testing purposes.